### ISSUE TITLE
[Bug]: Rest APIs not working with TLS enabled

### ISSUE THREAD

**Issue by nish112022** [milvus-io/milvus#36724]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:2.4.x
- Deployment mode(standalone or cluster): standalone
- MQ type(rocksmq, pulsar or kafka):    
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```

### Current Behavior

When I run this 
```
curl --request POST --url "http://localhost:19530/v2/vectordb/collections/describe"      --header "Authorization: Bearer 'root:milvus'"      --header "accept:
 application/json"      --header "content-type: application/json" -d '{
"dbName": "default",
"collectionName": "hello_milvus"
}'
```

It works, giving me the collection details.However, when I run the same command after enabling TLS mode=1 I get the error that **application/grpc expected not application json**

Link to thread:https://discord.com/channels/1160323594396635310/1293419285045182516

### Expected Behavior

The command should work

`curl --request POST --url "http://localhost:19531/v2/vectordb/collections/describe"      --header "Authorization: Bearer 'admin:3kMjCMzfL25w'"      --header "accept:
 application/json"      --header "content-type: application/json" 
 --cacert ./milvus-rest.cert -d '{
"dbName": "default",
"collectionName": "hello_milvus"
}'`

### Steps To Reproduce

```markdown
Enable TLSMode=1
Run Milvus
Give Curl command
```

**Comment by yanliang567** [comment]
/assign @haorenfsa 
does we need -secure in the request in this case? 
/unassign

**Comment by haorenfsa** [comment]
Hi @nish112022, thank you for the feedback. After tlsMode enabled, we should use  the scheme `https://` instead of `http://`, as the script below:
```shell
curl --request POST --url "https://localhost:19531/v2/vectordb/collections/describe"\
--header "Authorization: Bearer 'admin:3kMjCMzfL25w'"\
--header "accept: application/json"\
--header "content-type: application/json"\
--cacert ./milvus-rest.cert -d '{ "dbName": "default", "collectionName": "hello_milvus" }'
```

**Comment by nish112022** [comment]
MyBad @haorenfsa ,I should have explained a bit more.When I use the script you gave me , I get this error:
<img width="1135" alt="Screenshot 2024-10-10 at 4 24 10 PM" src="https://github.com/user-attachments/assets/3f6e84ce-d89b-409d-a63d-18952bb30485">

When I give the cert as -cacert field I get this error:**application/grpc expected not application json**
<img width="689" alt="Screenshot 2024-10-10 at 4 23 24 PM" src="https://github.com/user-attachments/assets/f76d0596-7a4b-4e9b-afd9-eb10c67273f0">

Below are the parameters I use for my hello_milvus.py.The same parameters should work here as well.
```
connections.connect(
  alias='defa',
  secure=True, server_name='localhost',
  server_pem_path='/root/milvus_2.4/milvus/configs/cert/server.pem',
  user='root',
  password='Milvus',
  host='localhost',
  port='19530',
)
```

**Comment by haorenfsa** [comment]
@nish112022  Oh, I get it.  It seems to be the issue of golang's mux server when serving https & gRPC over TLS on the same port.

For now you can walk around by configuring the http server to another port (for example, 8080)  in milvus configuration:
```yaml
proxy:
  http:
    port: 8080
```

**Comment by haorenfsa** [comment]
or you may also try force your client to use http/1.1, without changing configuration.
`curl --http1.1` ( same thing needs to be done with your http lib if you want to write code to access it later)
Since ALPN protocol decides which handler will handle the request, h2 for gRPC and http/1.1 for restful.
<img width="300" alt="image" src="https://github.com/user-attachments/assets/fd08e9e8-586b-435d-a42d-539f74962365">

**Comment by nish112022** [comment]
It seems forcing http1.1 doesn't work as well.To me it seems due to APLN. not being able to select the configuration correctly.
<img width="708" alt="Screenshot 2024-10-10 at 5 08 52 PM" src="https://github.com/user-attachments/assets/dc1e5d81-3fd9-4d3f-8ad2-ea8d91c7dc92">

**Comment by haorenfsa** [comment]
@nish112022 Yes, it's a bug in server side. by the way I just noticed that in the newest release,  @chyezh has forbidden enable tls when enable restful & grpc in the same port.

https://github.com/milvus-io/milvus/commit/e34fa0461b5dd26acbc5c84b63913cabc653c2e8#diff-6e86fc33d8695678bc197f60ce811f0a362cb3034cda0cc03823b5bc8822920eR75

I think we should fix this instead of forbidding the usage, in a way similar to this: https://ahmet.im/blog/grpc-http-mux-go/

**Comment by haorenfsa** [comment]
@nish112022 For now only this works for me. We'll fix this later. Thank you again for feedback.
```
proxy:
  http:
    port: 8080
```

**Comment by nish112022** [comment]
> @nish112022 Yes, it's a bug in server side. by the way I just noticed that in the newest release, @chyezh has forbidden enable tls when enable restful & grpc in the same port.
> 
> [e34fa04#diff-6e86fc33d8695678bc197f60ce811f0a362cb3034cda0cc03823b5bc8822920eR75](https://github.com/milvus-io/milvus/commit/e34fa0461b5dd26acbc5c84b63913cabc653c2e8#diff-6e86fc33d8695678bc197f60ce811f0a362cb3034cda0cc03823b5bc8822920eR75)
> 
> I think we should fix this instead of forbidding the usage, in a way similar to this: https://ahmet.im/blog/grpc-http-mux-go/

Yes, you are correct

**Comment by nish112022** [comment]
@haorenfsa May I know how this issue has been tackled in Zilliz Cloud?

**Comment by nish112022** [comment]
@haorenfsa So, I tried some approaches and there are essentially 2.
1.  There is a **property called as 	 NextProtos in here**, where in if I specify this particular order.It uses http1 for the rest endpoint and h2 for the grpc because of the order here.
tlsConfig := &tls.Config{
		Certificates: []tls.Certificate{cert},
		NextProtos:   []string{"http/1.1", "h2"}, // Enable HTTP/2 and HTTP/1.1
	}
	
	The **issue with this approach** is that it relies on differentiating requests based on the protocol (http/1.1 vs h2). If HTTP/1.1 is deprecated in the future, this logic may fail, as both gRPC and HTTP endpoints would need to share HTTP/2 (h2)

2. I tried the approach from the link you provided.I can separate the grpc and http requests even when both use h2 with tls. I will be attaching the code below.Let me know what do you think about this.I believe this should the approach we should take here.
[multiple_split.txt](https://github.com/user-attachments/files/18126084/multiple_split.txt)

**Comment by xiaofan-luan** [comment]
@nish112022 
feel free to submit a fix and we'd definitely like to take it if applicable

**Comment by ashkrisk** [comment]
/reopen
We are still planning to raise a patch for this
